src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTable.cs

```diff
@@ -0,0 +1,388 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using Microsoft.PowerFx.Core;
+using Microsoft.PowerFx.Core.App;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Binding.BindInfo;
+using Microsoft.PowerFx.Core.Entities;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Types.Enums;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx
+{
+    // Can pass to symbol table at compile time, and then it has a handle
+    // if we use it for execution at runtime. 
+    public interface ICanGetValue
+    {
+        FormulaValue Value { get; }
+    }
+
+    // NameInfo data to associate a symbol at bind time with a runtime config at runtime. 
+    internal class NameSymbol
+    {
+        public NameSymbol(string name)
+        {
+            Name = name;
+        }
+        
+        public string Name { get; private set; }
+    }
+
+    /// <summary>
+    /// Provides symbols to the engine. This includes variables (locals, globals), enums, options sets, and functions.
+    /// SymbolTables are mutable to support sessionful scenarios and can abe chained together. 
+    /// This is a publicly facing class around a <see cref="INameResolver"/>.
+    /// </summary>
+    [DebuggerDisplay("{DebugName}")]
+    public class SymbolTable : INameResolver, IGlobalSymbolNameResolver
+    {
+        // Changed on each update. 
+        // Host can use to ensure that a symbol table wasn't mutated on us.                 
+        private int _version = (int)DateTime.UtcNow.Ticks; // seed psuedo-random
+
+        public int VersionHash => _parent == null ? 
+            _version : 
+            (_version * -1521134295) + _parent.VersionHash;
```

Comment:
Unsecure and probably unsafe as two threads calling this code at the same time will produce the same result.

We could consider a fully secure option (but slow)

    using (var provider = new RNGCryptoServiceProvider()) 
    {
        var bytes = new byte[16];
        provider.GetBytes(bytes);

        return new Guid(bytes);
    }

or use a simple counter with Interlocked.Add with a static (very fast)

Reply:
These classes are not thread-safe - this is like the default BCL collections.  Host is responsible for locking if they're going to try and mutate from multiple threads.